File: markwon-inline-parser/src/main/java/io/noties/markwon/inlineparser/HtmlInlineProcessor.java
package io.noties.markwon.inlineparser;

import org.commonmark.node.HtmlInline;
import org.commonmark.node.Node;

import java.util.regex.Pattern;

public class HtmlInlineProcessor extends InlineProcessor {

    private static final String TAGNAME = "[A-Za-z][A-Za-z0-9-]*";
    private static final String ATTRIBUTENAME = "[a-zA-Z_:][a-zA-Z0-9:._-]*";
    private static final String UNQUOTEDVALUE = "[^\"'=<>`\\x00-\\x20]+";
    private static final String SINGLEQUOTEDVALUE = "'[^']*'";
    private static final String DOUBLEQUOTEDVALUE = "\"[^\"]*\"";
    private static final String ATTRIBUTEVALUE = "(?:" + UNQUOTEDVALUE + "|" + SINGLEQUOTEDVALUE + "|" + DOUBLEQUOTEDVALUE + ")";
    private static final String ATTRIBUTEVALUESPEC = "\\s*=\\s*" + ATTRIBUTEVALUE;
    private static final String ATTRIBUTE = "\\s+" + ATTRIBUTENAME + "(?:" + ATTRIBUTEVALUESPEC + ")?";

    private static final String OPENTAG = "<" + TAGNAME + "(?:" + ATTRIBUTE + ")*\\s*/?>";
    private static final String CLOSETAG = "</" + TAGNAME + "\\s*>";

    private static final String HTMLCOMMENT = "<!---->|<!--(?:-?[^>-])(?:-?[^-])*-->";
    private static final String PROCESSINGINSTRUCTION = "[<][?].*?[?][>]";
    private static final String DECLARATION = "<![A-Z]+\\s+[^>]*>";
    private static final String CDATA = "<!\\[CDATA\\[[\\s\\S]*?\\]\\]>";

    private static final String HTMLTAG = "(?:" + OPENTAG + "|" + CLOSETAG + "|" + HTMLCOMMENT
            + "|" + PROCESSINGINSTRUCTION + "|" + DECLARATION + "|" + CDATA + ")";
    private static final Pattern HTML_TAG = Pattern.compile("^" + HTMLTAG, Pattern.CASE_INSENSITIVE);

    @Override
    public char specialCharacter() {
        return '<';
    }

    @Override
    protected Node parse() {
        String matched = match(HTML_TAG);
        if (matched != null) {
            HtmlInline node = new HtmlInline();
            node.setLiteral(matched);
            return node;
        } else {
            return null;
        }
    }
}

File: markwon-inline-parser/src/main/java/io/noties/markwon/inlineparser/HtmlInlineProcessor.kt
package io.noties.markwon.inlineparser

import org.commonmark.node.HtmlInline
import org.commonmark.node.Node
import java.util.regex.Pattern

class HtmlInlineProcessor : InlineProcessor() {
    override fun specialCharacter(): Char {
        return '<'
    }

    override fun parse(): Node? {
        val matched = match(HTML_TAG)
        if (matched != null) {
            val node = HtmlInline()
            node.literal = matched
            return node
        } else {
            return null
        }
    }

    companion object {
        private const val TAGNAME = "[A-Za-z][A-Za-z0-9-]*"
        private const val ATTRIBUTENAME = "[a-zA-Z_:][a-zA-Z0-9:._-]*"
        private const val UNQUOTEDVALUE = "[^\"'=<>`\\x00-\\x20]+"
        private const val SINGLEQUOTEDVALUE = "'[^']*'"
        private const val DOUBLEQUOTEDVALUE = "\"[^\"]*\""
        private const val ATTRIBUTEVALUE = "(?:$UNQUOTEDVALUE|$SINGLEQUOTEDVALUE|$DOUBLEQUOTEDVALUE)"
        private const val ATTRIBUTEVALUESPEC = "\\s*=\\s*$ATTRIBUTEVALUE"
        private const val ATTRIBUTE = "\\s+$ATTRIBUTENAME(?:$ATTRIBUTEVALUESPEC)?"

        private const val OPENTAG = "<$TAGNAME(?:$ATTRIBUTE)*\\s*/?>"
        private const val CLOSETAG = "</$TAGNAME\\s*>"

        private const val HTMLCOMMENT = "<!---->|<!--(?:-?[^>-])(?:-?[^-])*-->"
        private const val PROCESSINGINSTRUCTION = "[<][?].*?[?][>]"
        private const val DECLARATION = "<![A-Z]+\\s+[^>]*>"
        private const val CDATA = "<!\\[CDATA\\[[\\s\\S]*?\\]\\]>"

        private const val HTMLTAG = ("(?:" + OPENTAG + "|" + CLOSETAG + "|" + HTMLCOMMENT
                + "|" + PROCESSINGINSTRUCTION + "|" + DECLARATION + "|" + CDATA + ")")
        private val HTML_TAG: Pattern = Pattern.compile("^$HTMLTAG", Pattern.CASE_INSENSITIVE)
    }
}